### PR TITLE
Feature/read_comsol_exports

### LINKED ISSUE #335: Feature: read comsol exports
## General

* pyfar version: 0.4.2

## Description

What do you think of a function to read files exported from Comsol (.txt or .csv), which automatically extracts the data incl. frequencies, labels etc.? Would it be a good function for pyfar.io?

### PR DESCRIPTION
**Pull request by ahms5**
### Which issue(s) are closed by this pull request?

Closes #335

### Changes proposed in this pull request:

- read comsol files with .txt, .dat and .csv
- support 'spreadsheet' format
- Frequency and Timedomain data 
- just a first draft
- output is a dict with meta data and the actual data as FreqData oder signal

## Project

**Do not assign your issue to a project.** This is done automatically.

### CODE REVIEW

**Comment by f-brinkmann** [comment]
We decided the following:
- parse the variables-name string to know how to reshape the data
- return an N-dimensional object that is correctly reshaped
- return dict with names of the dimensions as key and a list containing the values. E.g. {'theta0 deg': [0, 1, 2]}
- return positions and Coordinates object + list of simplices for meshes

**Comment by sikersten** [comment]
@mberz and I discussed that it probably makes sense to split functionality into 2 functions, something like

1. ``read_comsol(filename, expression, parameters=None...)`` just returns audio objects (TimeData or FrequencyData) (and coordinates?). It would be required to give at least one expression (but it could be a list or tuple). It would be nice to have the possibility to both return objects combining several expressions as well as separate objects for separate expressions. The channel axes correspond to the parameters (in case of simulations including parameter sweeps), which allows to correctly reshape.
2. ``read_comsol_header(filename)`` returns the meta data: _expressions_, _parameters_, _units_, _domain_, _time steps / frequency bins_. This functions allows to quickly inspect a file and lets you define the expression(s) ``read_comsol(...)`` is called with.

For example, consider a complex sound pressure as well as the real components of an intensity vector and two parameter sweeps for a source position written into a single file. This structure ensures that this data is not combined into a single (complex) numpy array, while at the same time let's you combine the 3D intensity data into one array. The resulting audio objects would have a 2D cshape, where the axes correspond to the to parameters defining the source positions.
Furthermore, 2 functions are easier to debug and restrict any ambiguity whenever actual data is read.

**Comment by f-brinkmann** [comment]
Sounds reasonable. Would it make sense to have only one function, that only returns the header if no expression is provided and returns the header + additional data otherwise?

**Comment by f-brinkmann** [comment]
We decided the following
- We return Coordinates if available
- We raise a warning if values are stored in dB
- Return order is: data, parameter values, coordinates

**Review by ahms5 — COMMENTED** [review]
test will fail since the bug in FrequencyData. See request #346
now it's final from my side, please review:
some comments:
- not sure how to deal with optional output probaly 
- could not test the docstring
- how to ignore warnings during testing

**Comment by sikersten** [comment]
The failing test for python 3.7. is caused by changes with numpy's `loadtxt` regarding the `converters` parameter. It was [updated in numpy 1.23](https://numpy.org/doc/stable/release/1.23.0-notes.html), but [python 3.7 was dropped](https://numpy.org/doc/stable/release/1.22.0-notes.html) since 1.22 .
I'd suggest to also drop the compatibility with python 3.7 with the next pyfar release. Any other ideas?

**Review by sikersten — APPROVED** [review]

**Comment by ahms5** [comment]
Implemented as discussed:
- loop through all variables for sorting the data
- include `expressions` and `parameters` parameter in `real_comsol`
- missing data will be nan and waring is raised
ready for review/comments

**Comment by ahms5** [comment]
@mberz and I decided to leave the mesh feature, it is not needed atm. It can be added later, if somebody need it.
Ready for review!

File: pyfar/io/io.py [hunk 1]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+
+
+def read_comsol_header(filename):
+    """Returns header information from exported data from COMSOL.
+
+    Parameters
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are '.txt', '.dat' and '.csv'. '.csv'
+        Files are strongly recommended, since .txt and .dat are inconvenient
```
Comment (f-brinkmann) [inline pyfar/io/io.py:631]: See above regarding the txt and dat files

File: pyfar/io/io.py [hunk 2]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    """Read exported data from COMSOL.
+
+    Parameters
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are '.txt', '.dat' and '.csv'. '.csv'
+        Files are strongly recommended, since .txt and .dat are inconvenient
+        in their format.
+    expressions : list[str], optional
+        This list of expressions defines the output expressions. An example
```
Comment (f-brinkmann) [inline pyfar/io/io.py:476]: - This is hard to understand for a non COMSOL user. Should we give at least a brief description of what expressions are in COMSOL?

- I would change the term 'defines the output expressions' because we are not returning expressions. How about 'defines which parts of the COMSOL data is returned as a pyfar audio object'?

- The links should start with pyfar.io but start with pyfar.dsp

- The link to `read_comsol_header` did not really clarify fro me. How about writing 'a list of all expressions can be obtained with ...'?

File: pyfar/io/io.py [hunk 3]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+        can be like ``expressions=['pabe.Lp_t']``. Also see
+        :py:func:`~pyfar.dsp.comsol_read_header`. By default, all expressions
+        get returned.
+    parameters : dict
+        This dict contains the parameters, which sould be read. An example can
+        be like ``parameters={'theta': [0.0, 0.7854], 'phi': [0., 1.5708]}``.
+        Also see :py:func:`~pyfar.dsp.comsol_read_header`. Default is None,
```
Comment (f-brinkmann) [inline pyfar/io/io.py:483]: the link should start with pyfar.io

File: pyfar/io/io.py [hunk 4]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+        Input file. Excepted input files are '.txt', '.dat' and '.csv'. '.csv'
+        Files are strongly recommended, since .txt and .dat are inconvenient
+        in their format.
+    expressions : list[str], optional
+        This list of expressions defines the output expressions. An example
+        can be like ``expressions=['pabe.Lp_t']``. Also see
+        :py:func:`~pyfar.dsp.comsol_read_header`. By default, all expressions
+        get returned.
+    parameters : dict
+        This dict contains the parameters, which sould be read. An example can
```
Comment (f-brinkmann) [inline pyfar/io/io.py:481]: Maybe a very brief description of what parameters are in COMSOL?

File: pyfar/io/io.py [hunk 5]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+        be like ``parameters={'theta': [0.0, 0.7854], 'phi': [0., 1.5708]}``.
+        Also see :py:func:`~pyfar.dsp.comsol_read_header`. Default is None,
+        then all parameters are read.
+
+    Returns
+    -------
+    data : Signal or FrequencyData
```
Comment (f-brinkmann) [inline pyfar/io/io.py:488]: Do we have to describe how the cshape of the audio object relates to `expression` and `parameters` or is that clear from the input data?

File: pyfar/io/io.py [hunk 6]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+        Returns a Signal or FrequencyData object depending on the input data.
+    coordinates : Coordinates, optional
+        A Coordinates Object based on the Input Data. If the input dimension is
```
Comment (f-brinkmann) [inline pyfar/io/io.py:491]: Does COMSOL always return cartesian coordinates? If yes, this should be noted. Otherwise using zero for missing data might lead to errors.

File: pyfar/io/io.py [hunk 7]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+        lower than three, the missing dimensions are set to zero. If there are
+        no positions given, this parameter will not be returned.
+
+    Raises
```
Comment (f-brinkmann) [inline pyfar/io/io.py:495]: So far, we did not include this section in our docstrings. Do we want that? I'm more in favor of verbose error messages that help the user to resolve the issue.

File: pyfar/io/io.py [hunk 8]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    Returns all expressions
+    >>> import pyfar as pf
+    >>> data, coordinates = pf.io.read_comsol('my_comsol_data.csv')
+    """
+
+    # check Datatype
+    suffix = pathlib.Path(filename).suffix
+    if suffix not in ['.txt', '.dat', '.csv']:
+        raise ValueError((
+            "Input path must be a .txt, .csv or .dat file"
```
Comment (f-brinkmann) [inline pyfar/io/io.py:514]: I would use 'input data' instead of 'input path'

File: pyfar/io/io.py [hunk 9]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+            f"but is of type {str(suffix)}"))
+
+    # get header
+    all_expressions, units, all_parameters, domain, domain_data \
+        = read_comsol_header(filename)
+    if 'dB' in units:
+        warnings.warn(r'Do you really want to use dB-values?')
```
Comment (f-brinkmann) [inline pyfar/io/io.py:521]: I would suggest a more verbose warning that tells the user why this might be risky. Maybe something like 'The data contains values in dB. Consider to de-logarithmize the data if possible. Otherwise any further processing of the data might lead to erroneous results.'

File: pyfar/io/io.py [hunk 10]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+                parameter_idxes = np.array(domain_idxes | True)
+                for key in parameters:
+                    parameter_idxes &= (parameter_header[key] == parameters_pairs[key][parameters_idx])
+                idxes = parameter_idxes & domain_idxes & expression_idxes
+                if any(idxes):
+                    data_out[:, i_expression, parameters_idx, i_domain] \
+                        = data_expressions[:, idxes].flatten()
+                else:
+                    data_inconsistent = True
+                    warnings.warn(r'Parameter data is inconsistent. Missing \
```
Comment (f-brinkmann) [inline pyfar/io/io.py:593]: Can we be more verbose as to what 'inconsistent' means.

File: pyfar/io/io.py [hunk 11]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are '.txt', '.dat' and '.csv'. '.csv'
+        Files are strongly recommended, since .txt and .dat are inconvenient
+        in their format.
+
+    Returns
+    -------
+    expressions : list[str]
+        This list of expressions defines the output expressions. An example
```
Comment (f-brinkmann) [inline pyfar/io/io.py:637]: As above, I would suggest to briefly instroduce the concepts of expressions and parameters.

File: pyfar/io/io.py [hunk 12]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+
+        ``'freq'``
+            The data are in frequency domain.
+        ``'time'``
+            The data are in time domain.
+    domain_data : list[]
+        This list contains the domain data from the input file. Depending
+        on the data in the input file, the output will be float or complex.
+
+    Raises
```
Comment (f-brinkmann) [inline pyfar/io/io.py:659]: See comment above on description of errors in docstrings

File: pyfar/io/io.py [hunk 13]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    ------
+    SyntaxError
+        Raises SyntaxError if the file format is not supported.
+    ValueError
+        Raises ValueError if the domain is not supported.
+
+    Examples
+    --------
+    Returns all expressions
```
Comment (f-brinkmann) [inline pyfar/io/io.py:668]: Should we include a print of the output in the example? Otherwise the example would only be a function call.

File: pyfar/io/io.py [hunk 14]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    >>> import pyfar as pf
+    >>> expressions, expressions_unit, parameters, domain, domain_data \
+    >>>     = pf.io.read_comsol_header('my_comsol_data.csv')
+    """
+    # Check Datatype
+    suffix = pathlib.Path(filename).suffix
+    if not suffix.endswith(('.txt', '.dat', '.csv')):
+        raise SyntaxError((
+            "Input path must be a .txt, .csv or .dat file"
```
Comment (f-brinkmann) [inline pyfar/io/io.py:677]: I would suggest 'data' instead of 'path'

File: tests/test_io.py [hunk 15]
```diff
@@ -642,3 +642,328 @@ def test_default_audio_subtype(default_audio_subtype_mock):
     subtype_return = pyfar.io.default_audio_subtype(format=audio_format)
     assert subtype_return == 'bla'
     default_audio_subtype_mock.assert_called_with(audio_format)
+
+
```
Comment (f-brinkmann) [inline tests/test_io.py:646]: How about moving the new tests to test_io_comsol.py to make them easier to locate?

File: pyfar/io/io.py [hunk 16]
```diff
@@ -461,6 +463,320 @@ def default_audio_subtype(format):
...
+    SyntaxError
+        Raises SyntaxError if the file format is not supported.
+    ValueError
+        Raises ValueError if the Frequency or Time data are inconvenient in the
+        input file.
+
+    Examples
+    --------
+    Returns all expressions
+    >>> import pyfar as pf
```
Comment (f-brinkmann) [inline pyfar/io/io.py:506]: I would suggest to add the call to `read_comsol_header` to the example otherwise the imported data would not be understandable, or would they?

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
Thanks for the effort. I mainly reviewed the docstrings and leave the code review to the comsol users. One general question

**Comment by ahms5** [comment]
applied suggestions by @sikersten and @f-brinkmann, ready for review by @mberz

File: pyfar/io/io.py [hunk 17]
```diff
@@ -490,6 +492,356 @@ def default_audio_subtype(format):
...
+    """Returns header information from exported data from COMSOL.
+
+    Parameters
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are '.txt', '.dat' and '.csv'. '.csv'
+        Files are strongly recommended, since .txt and .dat are varies in its
+        format definition.
+
+    Returns
```
Comment (sikersten) [inline pyfar/io/io.py:691]: These docstrings could be a little more detailed. What about explaining the meanings and corresponding options of `Expressions`, `Parameters`, `domain` and `domain_data` in Comsol first and and then describing the format of the returned data incl. the short examples?

File: pyfar/io/io.py [hunk 18]
```diff
@@ -490,6 +492,356 @@ def default_audio_subtype(format):
...
+        Returns the domain of the input data. Note that ``'lambda'`` domain
+        data is not allowed.
+
+        ``'freq'``
+            The data are in frequency domain.
+        ``'time'``
+            The data are in time domain.
+    domain_data : list[]
+        This list contains the domain data from the input file. Depending
+        on the data in the input file, the output will be float.
```
Comment (sikersten) [inline pyfar/io/io.py:717]:
```suggestion
    domain_data : list
        List containing the time samples or frequency bins depending
        on the data in the input file.
```

File: pyfar/io/io.py [hunk 19]
```diff
@@ -490,6 +492,356 @@ def default_audio_subtype(format):
...
+    >>> data, coordinates = pf.io.read_comsol(path, parameters=parameters)
+    >>> data
+    FrequencyData:
+    (8, 1, 3, 4) channels with 2 frequencies
+    >>> coordinates
+    1D Coordinates object with 8 points of cshape (8,)
+    domain: cart, convention: right, unit: met
+    coordinates: x in meters, y in meters, z in meters
+    Does not contain sampling weights
+    """
```
Comment (sikersten) [inline pyfar/io/io.py:568]: An example on how to remove the nans and reshape the audio object in case of `Specific combinations` in a `Parametric Sweep` would be nice.

**Review by sikersten — CHANGES_REQUESTED** [review]
Thanks for implementing! The functionality looks good to me, docstring comments only.

**Comment by ahms5** [comment]
all reviews included, thanks @sikersten and @f-brinkmann

File: pyfar/io/io.py [hunk 20]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    (8, 1, 4, 4) channels with 2 frequencies
+    >>> data.freq = data.freq.sum(where=~np.isnan(data.freq), axis=-2)
+    >>> data
+    FrequencyData:
+    (8, 1, 4) channels with 2 frequencies
+    >>> coordinates
+    1D Coordinates object with 8 points of cshape (8,)
+    domain: cart, convention: right, unit: met
+    coordinates: x in meters, y in meters, z in meters
+    Does not contain sampling weights
```
Comment (sikersten) [inline pyfar/io/io.py:603]:
```suggestion
    Assume a `Pressure Acoustics BEM Simulation` (named "pabe" in COMSOL) for
    two frequencies including a `Parametric Sweep` with `All Combinations` of
    certain incident angles theta and phi for the incident sound wave.
    The sound pressure ("p_t") is exported to a file `comsol_sample.csv`.

    Obtain information on the input file with
    :py:func:`~pyfar.io.read_comsol_header`.

    >>> expressions, units, parameters, domain, domain_data = \\
    >>>     pf.io.read_comsol_header('comsol_sample.csv')
    >>> expressions
    ['pabe.p_t']
    >>> units
    ['Pa']
    >>> parameters
    {'theta': [0.0, 0.7854, 1.5708, 2.3562, 3.1416],
     'phi': [0.0, 1.5708, 3.1416, 4.7124]}
    >>> domain
    'freq'
    >>> domain_data
    [100.0, 500.0]

    Read the data including all parameter combinations.

    >>> data, coordinates = pf.io.read_comsol('comsol_sample.csv')
    >>> data
    FrequencyData:
    (8, 1, 5, 4) channels with 2 frequencies
    >>> coordinates
    1D Coordinates object with 8 points of cshape (8,)
    domain: cart, convention: right, unit: met
    coordinates: x in meters, y in meters, z in meters
    Does not contain sampling weights

    Read the data with a subset of the parameters.

    >>> parameter_subset = {
            'theta': [1.5708, 2.3562, 3.1416],
            'phi': [0.0, 1.5708, 3.1416, 4.7124]}
    >>> data, coordinates = pf.io.read_comsol(
    >>>     'comsol_sample.csv', parameters=parameter_subset)
    >>> data
    FrequencyData:
    (8, 1, 3, 4) channels with 2 frequencies

    Now assume that instead of `All Combinations` only four `Specific
    Combinations` of the incident angles were simulated and exported to
    `comsol_sample_specific.csv`. Accordingly, the returned object contains
    nans. Remove the nans and squeeze `cshape` by summing of the parameter
    axes.

    >>> data, coordinates = pf.io.read_comsol('comsol_sample_specific.csv')
    UserWarning: Specific combinations is set in the Parametric Sweep in
    Comsol. Missing data is filled with nans.
    >>> data
    FrequencyData:
    (8, 1, 4, 4) channels with 2 frequencies
    >>> data.freq = data.freq.sum(where=~np.isnan(data.freq), axis=-2)
    >>> data
    FrequencyData:
    (8, 1, 4) channels with 2 frequencies
```

Comment (sikersten) [reply pyfar/io/io.py:603]: I'm not sure about the nan example, as it addtionally requires the parameter values in the combinations to be unique. Maybe the example is more confusing than helpful?

File: pyfar/io/io.py [hunk 21]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+        y = coords_data[:, 1] if n_dimension > 1 else np.zeros_like(x)
+        z = coords_data[:, 2] if n_dimension > 2 else np.zeros_like(x)
+        coordinates = Coordinates(x, y, z)
+    else:
+        coordinates = None
+    return data, coordinates
+
+
+def read_comsol_header(filename):
+    """Returns header information from exported data from COMSOL.
```
Comment (sikersten) [inline pyfar/io/io.py:723]:
```suggestion
    """Read header information on exported data from COMSOL Multiphysics.
```

File: pyfar/io/io.py [hunk 22]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    --------
+    Assuming a Pressure Acoustics BEM Simulation in COMSOL Multipysics. The
+    total sound pressure was exported. Accordingly the expression is
+    ``pabe.p_t`` with the Unit ``Pa``. Furthermore, incident angles and two
+    frequncy bins were set in a ``Parametric Sweep``. The incident angles are
+    defined in parameters and the frequeny bins (domain ``freq``) in
+    ``domain_data``. The output would be same for `All Combinations` and
+    `Specific Conbinations`.
+
+    >>> import pyfar as pf
```
Comment (sikersten) [inline pyfar/io/io.py:779]:
```suggestion
    Assume a `Pressure Acoustics BEM Simulation` (named "pabe" in COMSOL) for
    two frequencies 100 Hz and 500 Hz including a `Parametric Sweep` of certain
    incident angles theta and phi for the incident sound wave. The sound
    pressure ("p_t") is exported to a file `comsol_sample.csv`.

    Obtain information on the input file.

```

File: pyfar/io/io.py [hunk 23]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+        files are strongly recommended, since .txt and .dat-files vary in their
+        format definition.
+
+    Returns
+    -------
+    expressions : list of strings
+        The expressions in COMSOL are the pysical property of the calculated
+        solution, e.g. total sound pressure in Pa. The expression depends on
+        the physics and have unit. Multiple expressions are possible in one
+        file. An example can be like ``expressions=['pabe.p_t']``.
```
Comment (sikersten) [inline pyfar/io/io.py:746]:
```suggestion
        When exporting data in COMSOL, certain `Expressions` need to be
        specified that define the physical quantities to be exported. This
        function returns the expressions as a list, e.g.,
        ``expressions=['pabe.p_t']``.
```

File: pyfar/io/io.py [hunk 24]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    expressions_unit : list of strings
+        This list of expressions unit defines the units of the output
+        expressions. An example can be like ``expressions_unit=['Pa']``.
```
Comment (sikersten) [inline pyfar/io/io.py:749]:
```suggestion
    units : list of strings
        List of the units corresponding to `expressions`, e.g.,
        ``units=['Pa']``.
```

File: pyfar/io/io.py [hunk 25]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    parameters : dict
+        Solutions or expressions can be evalueted for different parameters,
+        such as differnet angles or other parameters. This dict contains all
+        parameters from the input file. If no parameters are available, an
+        emtpy dict is returnd. This dict is same for `All Combinations` and
+        `Specific Conbinations`. An example can be like
+        ``parameters={'theta': [0.0, 0.7854], 'phi': [0., 1.5708]}``.
```
Comment (sikersten) [inline pyfar/io/io.py:756]:
```suggestion
        COMSOL supports `Parametric Sweeps` to vary certain parameters in a
        sequence of simulations. This function returns the parameters
        and the parameter values as a dict. For example, in a study with a
        varying angle, the dict could be ``parameters={'theta': [0.0, 0.7854],
        'phi': [0., 1.5708]}``. If the input does not include parameters, an
        emtpy dict is returnd. Note that the dict is same for
        `All Combinations` and `Specific Combinations` of parameters as a
        distinction is not possible based due to the data format.
```

File: pyfar/io/io.py [hunk 26]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    domain : string
+        Returns the domain of the input data. Note that ``'lambda'`` domain
+        data is not allowed.
+
+        ``'freq'``
+            The data are in frequency domain.
+        ``'time'``
+            The data are in time domain.
```
Comment (sikersten) [inline pyfar/io/io.py:764]:
```suggestion
        Domain of the input data. Only time domain (``'time'``) or
        frequency domain (``'freq'``) simulations are supported.
```

File: pyfar/io/io.py [hunk 27]
```diff
@@ -490,6 +492,404 @@ def default_audio_subtype(format):
...
+    domain_data : list
+        List containing the time samples or frequency bins depending
+        on the data in the input file.
```
Comment (sikersten) [inline pyfar/io/io.py:767]:
```suggestion
        List containing the time samples or frequency bins depending
        on the domain of the data in the input file.
```

**Review by sikersten — CHANGES_REQUESTED** [review]
Mainly docstring suggestions :)

**Comment by mberz** [comment]
Building the documentation should have been fixed on develop, please merge develop into the feature branch.

File: pyfar/io/io.py [hunk 28]
```diff
@@ -491,6 +493,402 @@ def default_audio_subtype(format):
...
+        ``units=['Pa']``.
+    parameters : dict
+        COMSOL supports `Parametric Sweeps` to vary certain parameters in a
+        sequence of simulations. This function returns the parameters
+        and the parameter values as a dict. For example, in a study with a
+        varying angle, the dict could be ``parameters={'theta': [0.0, 0.7854],
+        'phi': [0., 1.5708]}``. If the input does not include parameters, an
+        emtpy dict is returnd. Note that the dict is same for
+        `All Combinations` and `Specific Combinations` of parameters as a
+        distinction is not possible based due to the data format.
```
Comment (sikersten) [inline pyfar/io/io.py:762]:
```suggestion
        distinction is not possible due to the data format.
```

File: pyfar/io/io.py [hunk 29]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+        as well as results for `Parametric Sweeps`.
+        For more information see
+        :py:func:`~pyfar.io.read_comsol_header`.
+
+    Parameters
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are .txt, .dat and .csv. .csv-
+        files are strongly recommended, since .txt and .dat-files vary in their
+        format definition.
```
Comment (mberz) [inline pyfar/io/io.py:512]:
```suggestion
        format definitions.
```

File: pyfar/io/io.py [hunk 30]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+    expressions : list of strings, optional
+        This parameter defines which parts of the COMSOL data is
+        returned as a pyfar FrequencyData or TimeData object. By default, all
+        expressions are returned. The COMSOL terminology is used, e.g.,
```
Comment (mberz) [inline pyfar/io/io.py:516]:
```suggestion
        expressions are returned. COMSOL terminology is used, e.g.,
```

File: pyfar/io/io.py [hunk 31]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+        ``expressions=['pabe.p_t']``.
+        Further information and a list of all expressions can be obtained with
+        :py:func:`~pyfar.io.read_comsol_header`.
+    parameters : dict, optional
+        COMSOL supports `Parametric Sweeps` to vary certain parameters in a
+        sequence of simulations. The `parameters` dict contains the parameters
+        and the parameter values that should be included in the returned data.
```
Comment (mberz) [inline pyfar/io/io.py:523]:
```suggestion
        and the their values that are to be returned.
```

File: pyfar/io/io.py [hunk 32]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+        Returns a TimeData or FrequencyData object depending the domain of the
+        input data. The output has the `cshape` (#points, #expressions,
+        #parameter1, #parameter2, ...). In case of missing parameter value
+        combinations in the input, the missing data is filled with nans.
+        If the input does not include parameters, the `cshape` is just
+        (#points, #expressions).
+    coordinates : Coordinates
+        Evaluation points extracted from the input file as Coordinates object.
+        The coordinate system is always cartesian. If the input dimension is
+        lower than three, missing dimensions are set to zero.
```
Comment (mberz) [inline pyfar/io/io.py:542]:
```suggestion
        lower than three, missing dimensions are filled with zeros.
```

File: pyfar/io/io.py [hunk 33]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+        needs to be `Spreadsheet`. This function supports several `Expressions`
+        as well as results for `Parametric Sweeps`.
+        For more information see below.
+
+    Parameters
+    ----------
+    filename : str, Path
+        Input file. Excepted input files are .txt, .dat and .csv. .csv-
+        files are strongly recommended, since .txt and .dat-files vary in their
+        format definition.
```
Comment (mberz) [inline pyfar/io/io.py:725]:
```suggestion
        format definitions.
```

File: pyfar/io/io.py [hunk 34]
```diff
@@ -491,6 +493,385 @@ def default_audio_subtype(format):
...
+        varying angle, the dict could be ``parameters={'theta': [0.0, 0.7854],
+        'phi': [0., 1.5708]}``. If the input does not include parameters, an
+        emtpy dict is returnd. Note that the dict is same for
+        `All Combinations` and `Specific Combinations` of parameters as a
+        distinction is not possible due to the data format.
+    domain : string
+        Domain of the input data. Only time domain (``'time'``) or
+        frequency domain (``'freq'``) simulations are supported.
+    domain_data : list
+        List containing the time samples or frequency bins depending
```
Comment (mberz) [inline pyfar/io/io.py:750]:
```suggestion
        List containing the sampling times or frequency bins depending
```

File: pyfar/io/io.py [hunk 35]
```diff
@@ -504,11 +885,11 @@ def _clipped_audio_subtypes():
         return
 
     collection = {}
-    signal = pyfar.Signal([-1.5, -1, -.5, 0, .5, 1, 1.5]*100, 44100)
+    signal = pyfar.Signal([-1.5, -1, -.5, 0, .5, 1, 1.5] * 100, 44100)
```
Comment (mberz) [inline pyfar/io/io.py:888]: These changes are not related to this PR, that's a couple of beers right there. 🍻

File: environment.yml [hunk 36]
```diff
@@ -4,7 +4,7 @@ channels:
   - conda-forge
 dependencies:
   - pip
-  - numpy>=1.14.0
+  - numpy>=1.23.0
```
Comment (mberz) [inline environment.yml:7]: That's only for mybinder.org.

**Review by mberz — CHANGES_REQUESTED** [review]
I'd say, from a user's perspective this looks good. I didn't not go through the code.

I only have docstring comments.

... and found some beers ;)
And, what's with the orientation binary which was deleted here?

File: pyfar/io/io.py (hunk 35 above)
Comment (ahms5) [reply pyfar/io/io.py:888]: Yes, you got me 🤦‍♀️ , but it was inviting me to do so 🙈

**Comment by ahms5** [comment]
made changes proposed by @mberz 
- beers ok, yes
- I reverted the deleted orientation file (but even I accidently deleted it, no test failed, not related to this PR)
